src/collapse/mod.rs

```diff
@@ -69,3 +85,183 @@ pub trait Collapse {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(map) => *map.entry(key).or_insert(0) += count,
+            MultiThreaded(arc) => {
+                arc.upsert(key, || count, |v| *v += count);
+            }
+        }
+    }
+
+    fn is_concurrent(&self) -> bool {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(_) => false,
+            MultiThreaded(_) => true,
+        }
+    }
+
+    fn write<W>(&mut self, mut writer: W) -> io::Result<()>
+    where
+        W: io::Write,
+    {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(ref mut map) => {
+                let mut contents: Vec<_> = map.drain().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+            MultiThreaded(ref mut arc) => {
+                let map = match Arc::get_mut(arc) {
+                    Some(map) => map,
+                    None => {
+                        return Err(io::Error::new(
+                            io::ErrorKind::Other,
+                            "Attempting to drain contents of a concurrent \
+                             hashmap when more than one thread has access to it, \
+                             which is not allowed.",
+                        ))
+                    }
+                };
+                let map = mem::replace(
+                    map,
+                    CHashMap::with_capacity(CAPACITY_MULTI_THREADED_HASHMAP),
+                );
+                let mut contents: Vec<_> = map.into_iter().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+struct Chunks<'a> {
+    current_index: usize,
+    data: &'a [u8],
+    indices: &'a [usize],
+}
+
+impl<'a> Iterator for Chunks<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let chunk = {
+            let start = self.indices.get(self.current_index)?;
+            match self.indices.get(self.current_index + 1) {
+                Some(end) => &self.data[*start..*end],
+                None => &self.data[*start..],
+            }
+        };
+        self.current_index += 1;
+        Some(chunk)
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/// Representation of input data that can be chunked up and shared
+/// across threads easily.
+struct Input {
+    /// Vector of byte indices representing the start of each chunk
+    indices: Vec<usize>,
+    /// The original data
+    inner: Vec<u8>,
+}
+
+impl Input {
+    fn new<F>(data: Vec<u8>, mut nthreads: usize, func: F) -> io::Result<Self>
```

Comment:
Here's a suggestion: rather than making a full pass on the input file first, what if we instead have a pool of worker threads and a channel to them that we send chunks to? This thing walks the file and sends off chunks of, say, 256 segments each to the pool as it scans the file. This would allow the work to start before we've finished the first pass over the file, and should give us even better speed-up, especially for large files.

Reply:
Any recs on the best threadpool library to use? I'm not up-to-date on "latest and greatest" for this in the rust ecosystem.

`tokio-threadpool` could work but is obviously "optimized" for IO tasks and requires a bunch of futures stuff (which I understand pretty well thanks to your streams :). Also, I bet you're pretty familiar with how it works. 

It'd be nice if we could continue to use scoped threads (since all our threads really need to be passed is a reference to the original input). This, of course, could be accomplished fairly easily with regular threads and unsafe code (the invariants of using unsafe to send a raw pointer across thread boundaries are not that hard to understand in our use case), but I'm assuming you would much prefer we stay in purely safe code, hence the thought that a threadpool of "scoped" threads would be preferable. 

Here's a crate that I've never used that I just came across while searching: [scoped_threadpool](https://crates.io/crates/scoped_threadpool). Heard of it? Know anything about it?

Is it hard to just make a threadpool struct from the crossbeam building blocks we're using now? (That would be the most fun)

Reply:
Nevermind. I think I know how to do this without any additional libraries.

Reply:
Yeah, I think I'd just do this with crossbeam :)

Reply:
So I've got a version that more or less does what's described in this comment (spins up a thread-pool and passes over the data just once, sending X stacks to the pool at a time). I think it will end up providing a speed up (particularly for large files).

Quick design question, though ... There are two versions that work, but each has it's trade-offs. I'm happy to go with whichever design you like best:

* **Option 1:** Spin up a threadpool everytime `collapse` is called and spin it down at the end of that function, the major assumption being that few people will need to call `collapse` on a `Folder` more than once in their code. This is simplest option and keeps the existing API, i.e. multi-threaded collapsing remains as simple as:
```rust
let reader = ...;
let writer = ...;
let options = Option::default();
let mut folder = Folder::from(options);
folder.collapse(reader, writer)?;
```

* **Option 2:** Keep a threadpool up and running for the lifetime of the `Folder` struct. This is strictly better in the sense that it eliminates work for anyone wanting to call `collapse` more than once (the threadpool is only created once), but it requires adding a lot of lifetime stuff to `Folder` since we're using scoped threads and would like to keep the ability to send just references across threads. This isn't hard to implement (again, I already have a working version), but it is significantly more cognitive overhead for anyone else coming along who wants to contribute a new collapser. The api (for multi-threaded code) would have to look something like this):
```rust
let reader: &[u8] = ...;
let writer = ...;
crossbeam::scope(|scope| {
    let mut folder = Folder::new(options, Some(scope));
    folder.collapse(reader, writer)
}).unwrap()?;
```

The single-threaded API could remain simple, e.g.
```rust
let reader = ...;
let writer = ...;
let mut folder = Folder::new(options, None);
folder.collapse(reader, writer)?;
```

And the `Folder` struct becomes something like:
```rust
pub struct Folder<'env, 'scope> {
    thread_pool: Option<ThreadPool<'scope, &'env [u8], Result<(), io::Error>>>,
    // all the other fields...
}
pub(crate) struct ThreadPool<'scope, Input, Output> {
    handles: Vec<ScopedJoinHandle<'scope, ()>>,
    receiver: channel::Receiver<Output>,
    sender: channel::Sender<Input>,
    sender_quit: channel::Sender<()>,
}
```

Reply:
I think I'm in favor of option 1 since the overhead of spawning a thread pool is likely miniscule compared to the cost of folding a file :)

Reply:
Had to take a break from this for a few days, but came back to it this morning. I have another design question for you. Choice between two different working alternatives:

* **What we were doing before**:
    * Allocate and read all data into one big vector upfront
    * Call read_line on big vector in a loop, writing data into a separate temporary vector that gets cleared on each iteration of the loop
    * **Summary:** Allocate 2 times / read data 2 times
* **Better option 1:** 
    * Do **not** read all data into one big buffer upfront
    * Call read_line on reader in a loop, 
    * Allocate one vector for each job sent to threadpool
    * **Summary:** Allocate many times / read data 1 time
    * **Pros/cons**: easy, 100% safe code, ~550 MiB/sec throughput on my machine (slower than option 2 below, but faster than what we were doing before)
* **Better option 2:** 
    * Allocate one big "buffer" upfront, where "buffer" is a little smarter than a simple vector. It acts exactly like a vector except when it runs out of capacity, instead of 1) allocating new space, 2) copying bytes to new space, and 3) freeing old space, it does **not** free old space. It keeps the old space around. 
    * Call read_line in a loop, writing into the "buffer" AND sending references to the buffer to the threads (this requires unsafe).
    * Is actually safe because 1) writing never frees old data as it could if we were using a vector and 2) we are only ever appending; so the mutations don't effect the read slices already sent across threads
    * 1 allocation / 1 read of data
    * a little more complicated, one line of unsafe code (writing to the "buffer" on each iteration of the loop), ~615 MiB/sec throughput on my machine, requires more memory in cases where we wouldn't know how big to make the big "buffer" upfront (e.g. where reading from stdio instead of a file, whose size can be known ahead of time). Memory requirements grow to ~2x size of input in worst case. (2^0 + 2^1 + 2^2 + ... + 2^N = 2^(N+1) - 1)

Reply:
The "Buffer"...

```rust
pub(crate) struct Buffer {
    inner: Vec<Vec<u8>>,
    index: usize,
}

impl Buffer {
    pub(crate) fn with_capacity(capacity: usize) -> Self {
        Self {
            inner: vec![Vec::with_capacity(capacity)],
            index: 0,
        }
    }

    fn current(&self) -> &Vec<u8> {
        &self.inner[self.index]
    }

    fn current_mut(&mut self) -> &mut Vec<u8> {
        &mut self.inner[self.index]
    }

    fn current_capacity(&self) -> usize {
        self.inner[self.index].capacity()
    }

    fn current_len(&self) -> usize {
        self.inner[self.index].len()
    }
}

impl Deref for Buffer {
    type Target = [u8];
    fn deref(&self) -> &Self::Target {
        self.current()
    }
}

/// `BufReadExt` provides an extra "read line" method that reads into a `&mut Buffer` instead of a
/// `&mut String`.
pub(crate) trait BufReadExt: io::BufRead {
    // Compare with standard library's `read_until` implementation here:
    // https://github.com/rust-lang/rust/blob/26ab32499c0114ae6e01e76374ae06bcd7a973bd/src/libstd/io/mod.rs#L1471
    fn read_line2(&mut self, buffer: &mut Buffer) -> io::Result<usize> {
        let mut read = 0;
        loop {
            let (done, used) = {
                let available = match self.fill_buf() {
                    Ok(n) => n,
                    Err(ref e) if e.kind() == io::ErrorKind::Interrupted => continue,
                    Err(e) => return Err(e),
                };
                let remaining = buffer.current_capacity() - buffer.current_len();
                match memchr::memchr(b'\n', available) {
                    Some(i) => {
                        if i + 1 > remaining {
                            // vector would reallocate! add a new one!
                            let mut new_vector = Vec::with_capacity(buffer.current_capacity() * 2);
                            new_vector.extend_from_slice(buffer.current());
                            buffer.inner.push(new_vector);
                            buffer.index += 1;
                        }
                        let current = buffer.current_mut();
                        current.extend_from_slice(&available[..=i]);
                        (true, i + 1)
                    }
                    None => {
                        if available.len() > remaining {
                            // vector would reallocate! add a new one!
                            let mut new_vector = Vec::with_capacity(buffer.current_capacity() * 2);
                            new_vector.extend_from_slice(buffer.current());
                            buffer.inner.push(new_vector);
                            buffer.index += 1;
                        }
                        let current = buffer.current_mut();
                        current.extend_from_slice(available);
                        (false, available.len())
                    }
                }
            };
            self.consume(used);
            read += used;
            if done || used == 0 {
                return Ok(read);
            }
        }
    }
}

impl<T> BufReadExt for T where T: io::BufRead {}
```

Reply:
Oh, and this might make it a little clearer as to why Option 2 needs unsafe:

```rust
let mut buf = Buffer::with_capacity(whatever);
let buf_ptr = &mut buf as *mut Buffer;
loop {
    let n = read.read_line2(unsafe { &mut *buf_ptr})?;
    // Do some stuff
    if at_point_where_need_to_send_job_to_pool {
        // Some stuff...
        sender.send(&buf[start...end]).unwrap();
    }
}

Reply:
I think I'm still in favor of Option 1. Those extra allocations should be pretty close to free, especially if we opt in to jemalloc again. Even without that, something like [`bumpalo`](https://crates.io/crates/bumpalo) should basically eliminate the cost of these allocations, and may be worth looking into as later improvements.